Support Egress IP on OVN-IC

tssurya opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-org/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Trivial changes are exempt from following this template.
If your change is non-trivial, please provide the following information:
-->

**- What this PR does and why is it needed**
This PR tracks the work required to support EIP on IC environments. Here are the following pieces that need to go in:

- [x] Split egress node allocator and egressIP CRD patching into new global egressIP controller in cluster-manager
- [x] Ensure healthcheck functions needed by egressSVC for non-IC is still kept in master - pkg/ovn - pieces are moved i.e there is slight code duplication for a while until we refactor egressSVC for IC and then we should be able to consolidate the healthcheck packages into one.
- [x] Keep the NBDB configuration bits in egressip zone-controller
- [x] Do SNATs and reroutes towards joinIP for non-IC environments as is today.
When IC is enabled:
- [x] Do SNATs towards egressIP only if egressNode is local to the zone
- [x] Do reroutes towards transit switch only if pod is local to the zone
- [x] Do static routes towards joinIP only if pod is non-local to the zone but egressnode is local to the zone
- [x] Do SNATs towards nodeIP only if pod is local to the zone.
- [x] Optimize database configurations on zone level to update only changed status items.
- [x] Split reconcileCloudPrivateIPConfig correctly (done to the best of my knowledge, we have moved over cloud IP stuff over to CM) - this part needs careful reviewing
- [ ] TODO: Can be part of new PR, delete static routes when node get's deleted from gatewayCleanUP code (also this might be unnecessary since delete events are processed in reverse order of add events for the handlers that share the same object type, in this case EgressNode and Node types. However the same luxury doesn't exist for races between pods and nodes, so might be worth doing this.)
- [ ] TODO: Add a sync function for the new LRSR similar to the ones we have for the other objects, this can be done in new PR.
- [x] TODO: Fix metrics for egressIP - new PR - https://github.com/ovn-org/ovn-kubernetes/pull/3611 

**- Special notes for reviewers**
This PR does not run a multi-zone lane, so if someone wants to see a CI signal see <https://github.com/ovn-org/ovn-kubernetes/pull/3561>. Please pay extra attention to cloud private IP parts since there is no CI signal here for that. I could be easily breaking things. We at least need to add unit tests for that beast somewhere...

**- How to verify it**

- [x] Fix existing tests by splitting allocator and patch bits into cluster-manager and keeping NBDB configuration checks in egressip zone controller
- [x] Add new unit tests on egressip zone controller side to test the reroutes and static routes -> this is mainly for IC enabled cluster environments.
- [ ] TODO: Test CloudPrivateIP reconcile properly. -> test this downstream since we don't have coverage here.
- [x] All existing e2e's are passing in non-IC and IC lanes.

**- Description for the changelog**
`Support egressIPs in Interconnect mode`

tssurya commented on go-controller/pkg/clustermanager/clustermanager.go:
```diff
@@ -254,6 +277,28 @@ func (cm *ClusterManager) StartClusterManager() error {
 	}
 
+	klog.Infof("%v", config.OVNKubernetesFeature.EnableEgressIP)
+	if config.OVNKubernetesFeature.EnableEgressIP {
```
Does this need to be moved to `Run` function actually? after `WatchNodes()` ? that looks like the spot to do it?

tssurya replied:
also update `hasResourceAnUpdateFunc`

tssurya replied:
move metrics as well!

flavio-fernandes commented:
my eyes, my eyes!!!!
😜

tssurya commented:
/retest-failed

tssurya commented:
TODO: Investigate https://github.com/ovn-org/ovn-kubernetes/actions/runs/4539786292/jobs/7999987381?pr=3386

tssurya commented:
/retest-failed

tssurya commented:
/retest-failed

tssurya commented on go-controller/pkg/ovn/default_network_controller.go:
```diff
@@ -741,22 +779,27 @@ func (h *defaultNetworkControllerEventHandler) AddResource(obj interface{}, from
 		node := obj.(*kapi.Node)
+		// Add the node to the tracker only if it belongs to the local zone.
+		klog.Infof("SURYA %v", config.OVNKubernetesFeature.EnableInterconnect)
+		if config.OVNKubernetesFeature.EnableInterconnect {
```
we can remove the feature flag because even in default deployments, the `isLocalZone` will mean all nodes are local to the zone.

tssurya commented on go-controller/pkg/ovn/default_network_controller.go:
```diff
@@ -741,22 +779,27 @@ func (h *defaultNetworkControllerEventHandler) AddResource(obj interface{}, from
+				h.oc.eIPC.localZoneNodes.Delete(node.Name)
+			}
+		}
 		if err := h.oc.setupNodeForEgress(node); err != nil {
```
we need to process nodes only if they are local to this zone, no need to care about other nodes.

tssurya commented on go-controller/pkg/ovn/default_network_controller.go:
```diff
@@ -851,7 +915,19 @@ func (h *defaultNetworkControllerEventHandler) UpdateResource(oldObj, newObj int
+		klog.Infof("SURYA %v", config.OVNKubernetesFeature.EnableInterconnect)
+		if config.OVNKubernetesFeature.EnableInterconnect {
+			// handle zone annotation update on the node
+			if h.oc.isLocalZoneNode(newNode) {
```
when a node moves zone, should we tear setup here and let the new master redo the setup? wondering how frequently a zone move can happen.

tssurya commented:
/retest-failed

tssurya commented:
/retest-failed

flavio-fernandes commented on dist/images/ovnkube.sh:
```diff
@@ -1252,6 +1252,17 @@ ovn-cluster-manager() {
+  if [[ -n "${ovn_egress_ip_healthcheck_port}" ]]; then
+      egressip_healthcheck_port_flag="--egressip-node-healthcheck-port=${ovn_egress_ip_healthcheck_port}"
+  fi
+  echo "egressip_flags: ${egressip_enabled_flag}, ${egressip_healthcheck_port_flag}"
```
👍

flavio-fernandes commented on go-controller/pkg/util/kube.go:
```diff
@@ -64,6 +64,8 @@ type OVNNodeClientset struct {
 
 type OVNClusterManagerClientset struct {
 	KubeClient            kubernetes.Interface
+	EgressIPClient        egressipclientset.Interface
```
👍

flavio-fernandes commented on go-controller/pkg/libovsdbops/router.go:
```diff
@@ -703,6 +703,36 @@ func DeleteLogicalRouterStaticRoutesWithPredicate(nbClient libovsdbclient.Client
+// DeleteLogicalRouterStaticRoutesWithPredicateOps looks up logical router static
+// routes from the cache based on a given predicate, and returns the ops to delete
+// them and remove them from the provided logical router
+func DeleteLogicalRouterStaticRoutesWithPredicateOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, routerName string, p logicalRouterStaticRoutePredicate) ([]libovsdb.Operation, error) {
```
how about changing `DeleteLogicalRouterStaticRoutesWithPredicate` to call this new function and then just 
```
	_, err = TransactAndCheck(nbClient, ops)
	return err
```

tssurya replied:
yup I had the exact same thoughts when I was writing it... yeah I could possible refactor the other one to call this instead :D

tssurya replied:
this is done, resolving.

flavio-fernandes reviewed (commented):
to be continued. the 4th commit is going to need my full attention and a fresh brain. ;)

flavio-fernandes commented:
hi @tssurya . While running unit tests on this PR, I seem to have hit a deadlock.
I will paste the whole output here, but this is the test where I saw it:
```
I0503 15:20:40.354376   14234 reflector.go:227] Stopping reflector *v1.Node (0s) from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.354104   14234 watch.go:183] Stopping fake watcher.
I0503 15:20:40.355536   14234 reflector.go:227] Stopping reflector *v1.EndpointSlice (0s) from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.353961   14234 reflector.go:227] Stopping reflector *v1.Service (0s) from k8s.io/client-go/informers/factory.go:150
[32m•[0m
[90m------------------------------[0m
[0mOVN Egress Service Operations[0m [90mon nodes changes[0m 
  [1mshould create/update/delete logical router policies, labels and annotations[0m
  [37m/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/egressservices_test.go:1344[0m
[1mSTEP[0m: creating two services with different selectors
I0503 15:20:40.373693   14234 reflector.go:221] Starting reflector *v1.NetworkPolicy (0s) from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.373923   14234 reflector.go:257] Listing and watching *v1.NetworkPolicy from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.374844   14234 reflector.go:221] Starting reflector *v1.Node (0s) from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.375063   14234 reflector.go:257] Listing and watching *v1.Node from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.376667   14234 reflector.go:221] Starting reflector *v1.Service (0s) from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.377172   14234 reflector.go:221] Starting reflector *v1.EndpointSlice (0s) from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.377576   14234 reflector.go:257] Listing and watching *v1.EndpointSlice from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.377658   14234 reflector.go:221] Starting reflector *v1.Pod (0s) from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.376983   14234 reflector.go:257] Listing and watching *v1.Service from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.378126   14234 reflector.go:257] Listing and watching *v1.Pod from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.379632   14234 reflector.go:221] Starting reflector *v1.Namespace (0s) from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.380050   14234 reflector.go:257] Listing and watching *v1.Namespace from k8s.io/client-go/informers/factory.go:150
I0503 15:20:40.477968   14234 shared_informer.go:303] caches populated
I0503 15:20:40.481113   14234 shared_informer.go:303] caches populated
I0503 15:20:40.483549   14234 shared_informer.go:303] caches populated
I0503 15:20:40.486266   14234 shared_informer.go:303] caches populated
I0503 15:20:40.487646   14234 shared_informer.go:303] caches populated
I0503 15:20:40.489975   14234 shared_informer.go:303] caches populated
panic: test timed out after 10m0s

goroutine 53874 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:2029 +0xbb
created by time.goFunc
	/usr/local/go/src/time/sleep.go:176 +0x48

goroutine 1 [chan receive, 9 minutes]:
testing.(*T).Run(0xc0004c31e0, {0x31374d4, 0xf}, 0x3279078)
[deadlock.txt.gz](https://github.com/ovn-org/ovn-kubernetes/files/11388774/deadlock.txt.gz)

	/usr/local/go/src/testing/testing.go:1487 +0x750
testing.runTests.func1(0x0?)
	/usr/local/go/src/testing/testing.go:1839 +0x9a
testing.tRunner(0xc0004c31e0, 0xc00028fba0)
	/usr/local/go/src/testing/testing.go:1439 +0x214
testing.runTests(0xc000576aa0?, {0x48dcf80, 0x9, 0x9}, {0x40?, 0x2e0f360?, 0x4910ea0?})
	/usr/local/go/src/testing/testing.go:1837 +0x7e5
testing.(*M).Run(0xc000576aa0)
	/usr/local/go/src/testing/testing.go:1719 +0xa72
main.main()
	_testmain.go:63 +0x2e5

goroutine 12226 [select]:
k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc0069d8660)
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/util/workqueue/delaying_queue.go:233 +0x48d
created by k8s.io/client-go/util/workqueue.newDelayingQueue
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/util/workqueue/delaying_queue.go:70 +0x3be

goroutine 3127 [chan receive, 9 minutes]:
github.com/ovn-org/libovsdb/client.(*ovsdbClient).watchForLeaderChange.func2()
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/ovn-org/libovsdb/client/client.go:1061 +0x66
created by github.com/ovn-org/libovsdb/client.(*ovsdbClient).watchForLeaderChange
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/ovn-org/libovsdb/client/client.go:1060 +0x65f

goroutine 25277 [chan receive]:
k8s.io/client-go/util/workqueue.(*Type).updateUnfinishedWorkLoop(0xc00b07b2c0)
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/util/workqueue/queue.go:271 +0x130
created by k8s.io/client-go/util/workqueue.newQueue
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/util/workqueue/queue.go:63 +0x34f

goroutine 2450 [select]:
k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc0047bef00)
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/util/workqueue/delaying_queue.go:233 +0x48d
created by k8s.io/client-go/util/workqueue.newDelayingQueue
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/util/workqueue/delaying_queue.go:70 +0x3be

goroutine 451 [chan receive, 9 minutes]:
github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).registerForInterrupts(0xc0004726e0, 0x322e322e322e3200?)
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/specrunner/spec_runner.go:223 +0xc5
created by github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/specrunner/spec_runner.go:60 +0x10f

goroutine 126 [chan receive, 9 minutes]:
github.com/ovn-org/libovsdb/client.(*ovsdbClient).watchForLeaderChange.func2()
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/ovn-org/libovsdb/client/client.go:1061 +0x66
created by github.com/ovn-org/libovsdb/client.(*ovsdbClient).watchForLeaderChange
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/ovn-org/libovsdb/client/client.go:1060 +0x65f

goroutine 90 [chan receive, 9 minutes]:
github.com/ovn-org/libovsdb/client.(*ovsdbClient).watchForLeaderChange.func2()
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/ovn-org/libovsdb/client/client.go:1061 +0x66
created by github.com/ovn-org/libovsdb/client.(*ovsdbClient).watchForLeaderChange
	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/ovn-org/libovsdb/client/client.go:1060 +0x65f
```
[deadlock.txt.gz](https://github.com/ovn-org/ovn-kubernetes/files/11388795/deadlock.txt.gz)

tssurya commented:
@flavio-fernandes : thanks for bringing this up, do you think its related to this PR? seems like a code path in the `watchForLeaderChange` + `should create/update/delete logical router policies, labels and annotations` ? I don't remember touching much of anything in `Egress Service` code actually.

tssurya commented:
> hi @tssurya . While running unit tests on this PR, I seem to have hit a deadlock. I will paste the whole output here, but this is the test where I saw it:
> 
> ```
> I0503 15:20:40.354376   14234 reflector.go:227] Stopping reflector *v1.Node (0s) from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.354104   14234 watch.go:183] Stopping fake watcher.
> I0503 15:20:40.355536   14234 reflector.go:227] Stopping reflector *v1.EndpointSlice (0s) from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.353961   14234 reflector.go:227] Stopping reflector *v1.Service (0s) from k8s.io/client-go/informers/factory.go:150
> �[32m•�[0m
> �[90m------------------------------�[0m
> �[0mOVN Egress Service Operations�[0m �[90mon nodes changes�[0m 
>   �[1mshould create/update/delete logical router policies, labels and annotations�[0m
>   �[37m/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/egressservices_test.go:1344�[0m
> �[1mSTEP�[0m: creating two services with different selectors
> I0503 15:20:40.373693   14234 reflector.go:221] Starting reflector *v1.NetworkPolicy (0s) from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.373923   14234 reflector.go:257] Listing and watching *v1.NetworkPolicy from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.374844   14234 reflector.go:221] Starting reflector *v1.Node (0s) from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.375063   14234 reflector.go:257] Listing and watching *v1.Node from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.376667   14234 reflector.go:221] Starting reflector *v1.Service (0s) from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.377172   14234 reflector.go:221] Starting reflector *v1.EndpointSlice (0s) from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.377576   14234 reflector.go:257] Listing and watching *v1.EndpointSlice from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.377658   14234 reflector.go:221] Starting reflector *v1.Pod (0s) from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.376983   14234 reflector.go:257] Listing and watching *v1.Service from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.378126   14234 reflector.go:257] Listing and watching *v1.Pod from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.379632   14234 reflector.go:221] Starting reflector *v1.Namespace (0s) from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.380050   14234 reflector.go:257] Listing and watching *v1.Namespace from k8s.io/client-go/informers/factory.go:150
> I0503 15:20:40.477968   14234 shared_informer.go:303] caches populated
> I0503 15:20:40.481113   14234 shared_informer.go:303] caches populated
> I0503 15:20:40.483549   14234 shared_informer.go:303] caches populated
> I0503 15:20:40.486266   14234 shared_informer.go:303] caches populated
> I0503 15:20:40.487646   14234 shared_informer.go:303] caches populated
> I0503 15:20:40.489975   14234 shared_informer.go:303] caches populated
> panic: test timed out after 10m0s
> 
> goroutine 53874 [running]:
> testing.(*M).startAlarm.func1()
> 	/usr/local/go/src/testing/testing.go:2029 +0xbb
> created by time.goFunc
> 	/usr/local/go/src/time/sleep.go:176 +0x48
> 
> goroutine 1 [chan receive, 9 minutes]:
> testing.(*T).Run(0xc0004c31e0, {0x31374d4, 0xf}, 0x3279078)
> [deadlock.txt.gz](https://github.com/ovn-org/ovn-kubernetes/files/11388774/deadlock.txt.gz)
> 
> 	/usr/local/go/src/testing/testing.go:1487 +0x750
> testing.runTests.func1(0x0?)
> 	/usr/local/go/src/testing/testing.go:1839 +0x9a
> testing.tRunner(0xc0004c31e0, 0xc00028fba0)
> 	/usr/local/go/src/testing/testing.go:1439 +0x214
> testing.runTests(0xc000576aa0?, {0x48dcf80, 0x9, 0x9}, {0x40?, 0x2e0f360?, 0x4910ea0?})
> 	/usr/local/go/src/testing/testing.go:1837 +0x7e5
> testing.(*M).Run(0xc000576aa0)
> 	/usr/local/go/src/testing/testing.go:1719 +0xa72
> main.main()
> 	_testmain.go:63 +0x2e5
> 
> goroutine 12226 [select]:
> k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc0069d8660)
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/util/workqueue/delaying_queue.go:233 +0x48d
> created by k8s.io/client-go/util/workqueue.newDelayingQueue
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/util/workqueue/delaying_queue.go:70 +0x3be
> 
> goroutine 3127 [chan receive, 9 minutes]:
> github.com/ovn-org/libovsdb/client.(*ovsdbClient).watchForLeaderChange.func2()
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/ovn-org/libovsdb/client/client.go:1061 +0x66
> created by github.com/ovn-org/libovsdb/client.(*ovsdbClient).watchForLeaderChange
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/ovn-org/libovsdb/client/client.go:1060 +0x65f
> 
> goroutine 25277 [chan receive]:
> k8s.io/client-go/util/workqueue.(*Type).updateUnfinishedWorkLoop(0xc00b07b2c0)
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/util/workqueue/queue.go:271 +0x130
> created by k8s.io/client-go/util/workqueue.newQueue
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/util/workqueue/queue.go:63 +0x34f
> 
> goroutine 2450 [select]:
> k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc0047bef00)
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/util/workqueue/delaying_queue.go:233 +0x48d
> created by k8s.io/client-go/util/workqueue.newDelayingQueue
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/util/workqueue/delaying_queue.go:70 +0x3be
> 
> goroutine 451 [chan receive, 9 minutes]:
> github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).registerForInterrupts(0xc0004726e0, 0x322e322e322e3200?)
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/specrunner/spec_runner.go:223 +0xc5
> created by github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/specrunner/spec_runner.go:60 +0x10f
> 
> goroutine 126 [chan receive, 9 minutes]:
> github.com/ovn-org/libovsdb/client.(*ovsdbClient).watchForLeaderChange.func2()
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/ovn-org/libovsdb/client/client.go:1061 +0x66
> created by github.com/ovn-org/libovsdb/client.(*ovsdbClient).watchForLeaderChange
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/ovn-org/libovsdb/client/client.go:1060 +0x65f
> 
> goroutine 90 [chan receive, 9 minutes]:
> github.com/ovn-org/libovsdb/client.(*ovsdbClient).watchForLeaderChange.func2()
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/ovn-org/libovsdb/client/client.go:1061 +0x66
> created by github.com/ovn-org/libovsdb/client.(*ovsdbClient).watchForLeaderChange
> 	/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/ovn-org/libovsdb/client/client.go:1060 +0x65f
> ```
> 
> [deadlock.txt.gz](https://github.com/ovn-org/ovn-kubernetes/files/11388795/deadlock.txt.gz)

maybe we open a new issue to track this? Can't see how my PR is related to this...

flavio-fernandes commented:
@tssurya : short of using ` export OVN_INTERCONNECT_ENABLE=true` with PR https://github.com/ovn-org/ovn-kubernetes/pull/3561, do you see any value in trying this PR in kind deployments using `OVN_SEPARATE_CLUSTER_MANAGER=true` environment set?

flavio-fernandes commented on test/e2e/egressip.go:
```diff
@@ -1037,19 +1037,27 @@ spec:
 		ginkgo.By("7. Check the OVN DB to ensure no SNATs are added for the standby egressIP")
 		dbPods, err := framework.RunKubectl("ovn-kubernetes", "get", "pods", "-l", "name=ovnkube-db", "-o=jsonpath='{.items..metadata.name}'")
+		dbContainerName := "nb-ovsdb"
+		if isInterconnectEnabled() {
```
Assuming Numan keeps the container name the same for the db, independent of `isInterconnetEnabled`, we will not need this, right?

tssurya replied:
right. we will still need to pick up that specific node we care about but container names should not change.

flavio-fernandes reviewed (commented):
more comments; to be continued... :)

flavio-fernandes commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -2510,52 +2015,80 @@ func (e *egressIPController) deleteReroutePolicyOps(ops []ovsdb.Operation, podIP
+		routerName := util.GetGatewayRouterFromNode(status.Node)
+		natPred := func(nat *nbdb.NAT) bool {
+			// TODO(tssurya): investigate if we should be deleting NATs only from status.Node
+			return nat.ExternalIDs["name"] == name && nat.ExternalIP == status.EgressIP
```
oh, very good point. In this new world, we may be more susceptible to have mid air collisions where node X is removing and node Y is adding the nat for the given eip due to a revalidation. But since they are distinct NB dbs, it may be okay, agree?

tssurya replied:
yes +1 in IC world we should be good since DBs will be distinct.

flavio-fernandes commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1609,18 +1259,30 @@ func (oc *DefaultNetworkController) generateCacheForEgressIP() (map[string]egres
-				klog.Errorf("Unable to retrieve gateway IP for node: %s, protocol is IPv6: %v, err: %v", status.Node, isEgressIPv6, err)
-				continue
+			_, ok := oc.localZoneNodes.Load(status.Node)
+			if !config.OVNKubernetesFeature.EnableInterconnect || ok {
```
if localZones was always accurate, we could just:
```
if _, isLocalZoneNode := oc.localZoneNodes.Load(status.Node) ; isLocalZoneNode {
...
}
```
wdyt?

tssurya replied:
this is related to the previous comment right ? -> https://github.com/ovn-org/ovn-kubernetes/pull/3386/files#r1186521805 ? What you mean here is to remove the bool check for whether the feature is enabled or not?

flavio-fernandes replied:
yup, sorry for the repeat. I think you answered that above.

flavio-fernandes commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1609,18 +1259,30 @@ func (oc *DefaultNetworkController) generateCacheForEgressIP() (map[string]egres
+			} else {
+				nextHopIP, err = oc.eIPC.getTransitSwitchIP(status.Node)
+				if err != nil {
+					return nil, fmt.Errorf("unable to fetch transit switch IP for node %s: %v", status.Node, err)
```
Any reason why failing to find join is a `continue`, while transit is a `return`?

tssurya replied:
yeah good point. its because earlier we were continuing and then I thought should we really continue here or return ? I am torn between strongly faling versus silently ignoring.. I guess continue is the right approach because there will be a brief period of time when probably node's annotations are not yet added? I'll stick with what we were doing before unless someone feels otherwise.

flavio-fernandes replied:
ack. my suggestion would be to put a comment here explaining that reasoning. it would be good to be consistent in how to handle this.

tssurya replied:
yes I changed this into using continue as well so that its like the other parts of the code...

flavio-fernandes commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1993,7 +1459,7 @@ func (oc *DefaultNetworkController) deleteEgressNode(nodeName string) error {
 				// difference, hence: call the reconciliation function directly.
 				if err := oc.reconcileEgressIP(nil, &egressIP); err != nil {
-					errorAggregate = append(errorAggregate, fmt.Errorf("Re-assignment for EgressIP: %s failed, unable to update object, err: %v", egressIP.Name, err))
+					errorAggregate = append(errorAggregate, fmt.Errorf("re-assignment for EgressIP: %s failed, unable to update object, err: %v", egressIP.Name, err))
```
since you are changing this, could you also `%w`, pls? :)

flavio-fernandes commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -2225,6 +1614,8 @@ type egressIPController struct {
 	// EgressIP Node reachability gRPC port (0 means it should use dial instead)
 	egressIPNodeHealthCheckPort int
+	// A cache that maintains the list of nodes that belong to this zone
+	localZoneNodes sync.Map
```
👍

flavio-fernandes commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -2480,23 +1924,84 @@ func (e *egressIPController) createReroutePolicyOps(ops []ovsdb.Operation, podIP
+			IPPrefix: podIPNet.IP.String(),
+			Nexthop:  gatewayRouterIP.String(),
+			ExternalIDs: map[string]string{
+				"name": egressIPName,
```
assuming this is a brand new static route we will be introducing in the codebase going fwd, could we add some additional external-ids here? Something explicit like **"owner": "hello-world-I-am-an-egress-ip-static-route"**  
but better than that, of course :) I ask because having something like that makes the syncStaticRoutes a lot safer to prune stale stuff. And also a todo placeholder here to ensure that sync is able to remove/fix these static routes!

tssurya replied:
hmm do we need another externalID in addition to the egressIPName one? technically speaking that is the owner and we use that for deleting stuff like we do for other objects..

flavio-fernandes replied:
My thinking is that tables like static routes are shared amongst different functionalities on ovnk8. Reconciling it may be a little tricky. But I do agree that managing more things can be a possibility of getting more bugs down the road.
Cost-wise, I think that another key:value is not a big deal. The secondary indexing functionality makes it super fast too. Let's see what others think about that and see if this is a good idea or a silly idea. ;)

trozet replied:
we should use unique ids with client side indexes like @jcaamano and @npinaeva added for other resource types. ACL example: https://github.com/ovn-org/ovn-kubernetes/pull/3521

Then you dont need to do predicate searches on source ip, next hop, etc, we just use the key and do fast lookup. Not sure this is in scope for this PR though as it would probably require also adding db indexes for all other static routes that we add. Also, see my earlier comment about why do we even use static routes.

tssurya replied:
I agree we should used indexes if there is perf gain. That will need to be a new epic since eIP code is so complex, don't want to couple with IC rn.

flavio-fernandes commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -0,0 +1,1751 @@
+	// we restart and then have these handlers act on stale data when they
+	// sync.
+	if eIPC.egressNodeHandler, err = eIPC.WatchEgressNodes(); err != nil {
+		return fmt.Errorf("unable to watch egress nodes %v", err)
```
`%w`

tssurya replied:
done!

flavio-fernandes reviewed (commented):
some more comments. I will need one more pass on this commit because my eyes are getting crossed at this point. ;)
all in all, this looks really good, @tssurya !

tssurya commented:
thanks for the review @flavio-fernandes : working on resolving them... but keep them coming hehe !

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -56,256 +50,109 @@ func getEgressIPAddrSetDbIDs(name egressIpAddrSetName, controller string) *libov
+		//        2) need setup
+		//        3) need no-op
+		if !reflect.DeepEqual(oldEIP.Status.Items, newEIP.Status.Items) {
+			statusToRemove := make(map[string]egressipv1.EgressIPStatusItem, 0)
```
hmm I actually took these names from the legacy egressIP code.. we try to compute what to keep and remove and then do the actual delete and add accordingly. If I rename to "Process" its not clear w.r.t to Keep part I feel.

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -56,256 +50,109 @@ func getEgressIPAddrSetDbIDs(name egressIpAddrSetName, controller string) *libov
-	// everything for all pods which match the new.
-	if len(ipsToAssign) == 0 &&
-		len(statusToRemove) == 0 {
+		oldNamespaceSelector, err := metav1.LabelSelectorAsSelector(&oldEIP.Spec.NamespaceSelector)
```
done.

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1095,27 +705,50 @@ func (oc *DefaultNetworkController) addPodEgressIPAssignments(name string, statu
 	if len(statusAssignments) == 0 {
 		return nil
 	}
+	if config.OVNKubernetesFeature.EnableInterconnect {
```
yes technically speaking this should work fine... but the reason why I did this was all the logic I have within the whole add and delete of egressIPs... I check for the either IC isn't enabled OR its enabled and present in cache in all parts of the code.. but I could do away with those checks and always do this like you said.. then its a matter of whether its present in cache or not. If we don't care about the cache performance here I can do what you are suggesting. I just wanted to keep non-IC as close as possible to what we used to have...

tssurya replied:
I have changed code to incorporate this suggestion, so far things are looking good...unit tests pass.. let's see how this fair's in e2e tests..

tssurya replied:
I have fixed this, seems to work fine..

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1380,6 +994,24 @@ func (oc *DefaultNetworkController) syncEgressIPs(namespaces []interface{}) erro
+		// NOTE: Even at this stage, there can be race; the bnc.zone might be the nodeName
+		// while the node's annotations are not yet set, so it still shows global.
+		// The EgressNode updates should constantly update this cache as nodes get added, updated and removed
+		if oc.isLocalZoneNode(node) {
```
oh yeah I could just store what I get out of this.. true. done.

tssurya commented on go-controller/pkg/ovn/egressip_test.go:
```diff
@@ -263,79 +151,313 @@ var _ = ginkgo.Describe("OVN master EgressIP Operations", func() {
+					gomega.Eventually(lsp.Options["nat-addresses"]).Should(gomega.Equal("router"))
+					gomega.Eventually(lsp.Options["exclude-lb-vips-from-garp"]).Should(gomega.Equal("true"))
+
+					// NOTE: Cluster manager is the one who patches the egressIP object.
```
yeah, good suggestion. let me do that!

tssurya replied:
this is done.

tssurya commented on test/e2e/egressip.go:
```diff
@@ -1037,19 +1037,27 @@ spec:
 
 		ginkgo.By("7. Check the OVN DB to ensure no SNATs are added for the standby egressIP")
 		dbPods, err := framework.RunKubectl("ovn-kubernetes", "get", "pods", "-l", "name=ovnkube-db", "-o=jsonpath='{.items..metadata.name}'")
+		dbContainerName := "nb-ovsdb"
```
ack yeah maybe that's something we want to fix in a separate PR, I don't know what is a good timeout here - wonder if we are hitting this in the CI as well?

flavio-fernandes commented on test/e2e/util.go:
```diff
@@ -1061,3 +1061,8 @@ func randStr(n int) string {
 	return string(b)
 }
+
+func isInterconnectEnabled() bool {
```
Something as simple as looking to see if the `transit switch exists` could be a way of finding out.
ack. I was thinking we could get some hints on the ovn-ic from the ovn db itself. 

`ovn-nbctl ls-list | grep -i transit | wc -l`

But no biggie; we can leave this as it is for now if this turn out to be too tricky.

tssurya replied:
ack that can also work, but I'd rather do this the env way than looking into the DB... for now I'll leave this as is, let's see what others say.

flavio-fernandes commented:
@tssurya I spent a few hours running this PR on a kind cluster with env `OVN_SEPARATE_CLUSTER_MANAGER=true` and closely 
- watch the logs from `ovnkube-cluster-manager` container 
- monitor ovn dbs
after creating eips and a deployment where I could manually:
- scale the number of pods
- modify the label `ovn-worker k8s.ovn.org/egress-assignable`
- block/allow traffic to port 9107
to force eip revalidations. While that does not filly exercise ovn-ic where nodes reach each other via the transit switch, it did exercise`ovnkube-cluster-manager` and `ovnkube-network-controller-manager` container interactions.
All good there! I'm curious to see that @kyrtapz, @trozet  and @jcaamano 's eagle eyes catch what I don't see, but this PR is a /lgtm as of now. Good stuff!!!

```
❯ docker exec -it ovn-worker bash                                                                                        │
root@ovn-worker:/# /sbin/iptables -A OUTPUT -p tcp --sport 9107 -j DROP                                                  │
root@ovn-worker:/# /sbin/iptables -D OUTPUT -p tcp --sport 9107 -j DROP                                                  │
root@ovn-worker:/#


❯ POD=$(kubectl get pod -n ovn-kubernetes -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}{end}' | grep ovnkube-db- | head -1) && \
  kubectl exec -ti $POD -n ovn-kubernetes -c nb-ovsdb -- bash || { echo "Failed: $?" }
[root@ovn-control-plane ~]# watch -d -n 3 "ovn-nbctl lr-policy-list ovn_cluster_router | grep '100 ' ; ovn-nbctl --format=csv --no-heading find nat external_ids:name=egressip"

❯ alias kn
kn='kubectl -n ovn-kubernetes'
❯ kn logs ovnkube-master-576c97b94d-dwkfk -c ovnkube-cluster-manager --follow
❯ kn logs ovnkube-master-576c97b94d-dwkfk -c ovnkube-network-controller-manager --follow

kubectl label namespaces default name=helloegress
kubectl label nodes ovn-worker k8s.ovn.org/egress-assignable=dummy
kubectl label nodes ovn-worker2 k8s.ovn.org/egress-assignable=dummy

kubectl label namespaces default name-
kubectl label nodes ovn-worker k8s.ovn.org/egress-assignable-
kubectl label nodes ovn-worker2 k8s.ovn.org/egress-assignable-
```

```
❯ cat eip.yaml
apiVersion: k8s.ovn.org/v1
kind: EgressIP
metadata:
    name: egressip
spec:
    egressIPs:
    - 172.18.1.4
    namespaceSelector:
        matchLabels:
            name: helloegress

❯ cat eip_2_nh.yaml
apiVersion: k8s.ovn.org/v1
kind: EgressIP
metadata:
    name: egressip
spec:
    egressIPs:
    - 172.18.1.4
    - 172.18.1.5
    namespaceSelector:
        matchLabels:
            name: helloegress

❯ cat eip_2_nh_podSelector.yaml
apiVersion: k8s.ovn.org/v1
kind: EgressIP
metadata:
    name: egressip
spec:
    egressIPs:
    - 172.18.1.4
    - 172.18.1.5
    namespaceSelector: {}
    podSelector:
        matchLabels:
            wants: egress
```

```
❯ cat deploy.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: deep
spec:
  replicas: 2
  selector:
    matchLabels:
      wants: egress
  template:
    metadata:
      labels:
        wants: egress
    spec:
      containers:
      - name: egressip-pod-container
        image: k8s.gcr.io/e2e-test-images/agnhost:2.26
        imagePullPolicy: IfNotPresent
        command:
        - sh
        - "-c"
        - "set -xe\n
          \ echo Hello Kubernetes!\n
          \ trap : TERM INT; sleep infinity & wait"

❯ k scale deployment deep --replicas 50
❯ k scale deployment deep --replicas 1
❯ k apply -f eip_2_nh_podSelector.yaml
❯ k apply -f eip_2_nh.yaml
❯ k apply -f eip.yaml
```

flavio-fernandes reviewed (commented):
All looks good here. Next, I will look at this code under PR https://github.com/ovn-org/ovn-kubernetes/pull/3561 to see the static routes and transit switch in action.

tssurya commented:
> @tssurya : short of using ` export OVN_INTERCONNECT_ENABLE=true` with PR #3561, do you see any value in trying this PR in kind deployments using `OVN_SEPARATE_CLUSTER_MANAGER=true` environment set?

hmm good question, I think no.. we will always be interested in INTERCONNECT_ENABLE=true and this feature is more for d/s than u/s.

tssurya commented:
> UGH. Conflict because of #3546... fixing that as well.
> 
> * spoke to @kyrtapz on some of the cloudPrivateEgressIP parts... let's see if we can remove privateIP watch from master totally.

done! removed this and also changed the name to `nodeZoneState` as discussed.

tssurya commented:
@kyrtapz @trozet PTAL

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -0,0 +1,1768 @@
+			// trigger an event for the ovnkube-master to take action upon.
+			// Note that once we figure out the statusToAdd parts below we will trigger an
+			// update to cloudPrivateIP object which will trigger another patch for the eIP object.
+			if err := eIPC.patchReplaceEgressIPStatus(name, statusToKeep); err != nil {
```
@kyrtapz : It turned out to be quiet simple..... though I had to double check what each variable was doing.. does this make sense to you?
from my understanding:

1. `status` has the fresh set of statuses from egressIP spec
2. `statusToKeep` at this stage has valid statuses -> `status` minus `invalidStatus`
3. `statusToRemove` at this stage has invalid statuses -> `status` minus `validStatus`

At this point we don't need to consider ipsToRemove and ipsToAdd I think? because that will be done after `executeCloudPrivateIPConfigChange` ? Here we only care about ensuring we are removing the invalid ones... Based on that the logic I took was to update the statusToKeep. WDYT?

kyrtapz replied:
Makes sense to me.
In the previous code this is how we handled statusToRemove for cloud:
```
if len(statusToRemove) > 0 {
	// Delete all assignments that are to be removed from the allocator
	// cache. If we don't do this we will occupy assignment positions for
	// the ipsToAdd, even though statusToRemove will be removed afterwards
	oc.deleteAllocatorEgressIPAssignments(statusToRemove)
	if err := oc.deleteEgressIPAssignments(name, statusToRemove); err != nil {
		return err
	}
}
...
if err := oc.executeCloudPrivateIPConfigChange(name, statusToAdd, statusToRemove); err != nil {
	return err
}
```
To achieve something similar we need to update the EIP status so it doesn't include statusToRemove.

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -0,0 +1,1768 @@
+			staleEgressIPs.Delete(toRemove.EgressIP)
+		}
+		for staleEgressIP := range staleEgressIPs {
+			if nodeName := eIPC.deleteAllocatorEgressIPAssignmentIfExists(name, staleEgressIP); nodeName != "" {
```
@kyrtapz and as for this case, I checked on master code, it seems we weren't called delete setup for NBDB for the stale ones so I don't think I need to do any patch replace here...

kyrtapz commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -0,0 +1,1768 @@
+	// or kept.
+	status := []egressipv1.EgressIPStatusItem{}
+
+	// Initialize two empty objects as to avoid SIGSEGV. The code should play
```
NIT: I find this comment confusing, there is only one object initialized here?

tssurya replied:
yea thanks again, i SHOULD have read each line.. and ensured comments didn't go stale. let me try and do that at least across the reconcile function.

kyrtapz commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -0,0 +1,1768 @@
+		ipsToRemove.Insert(status.EgressIP)
+	}
+	if ipsToRemove.Len() > 0 {
+		// The following is added as to ensure that we only add after having
```
I know this comment is copied but please rephrase it so it reflects the new behavior. egressIPController doesn't configure the DBs.

tssurya replied:
I think it is still relevant to keep the OVN DB parts, just that we need to remember this will be done asynchronously.. I will clarify this up.

kyrtapz reviewed (commented):
I focused on CNCC integration in clustermanager and the general logic looks good to me.

tssurya commented:
> UGH. Conflict because of #3546... fixing that as well.
> 
>

UGH. AGAIN Conflict because of https://github.com/ovn-org/ovn-kubernetes/pull/3585.... need to fix that as well.

trozet commented on go-controller/pkg/clustermanager/egressip_controller_test.go:
```diff
@@ -0,0 +1,2349 @@
+package clustermanager
+
```
what tests in this file are new and not copied over?

tssurya replied:
All of them were copied over. Nothing new in this file. New UTs were added only in `egressip_test.go` in `ovn` pkg to reflect the new its for LRSR and the way we add and remove things in NBDB.

trozet reviewed (changes requested):
reviewed up to half of go-controller/pkg/ovn/default_network_controller.go

will finish tmrw

trozet reviewed (changes requested):
reviewed up to half of go-controller/pkg/ovn/default_network_controller.go

will finish tmrw

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -2440,6 +1612,19 @@ func ipFamilyName(isIPv6 bool) string {
+	if err != nil {
+		return "", fmt.Errorf("unable to fetch transit switch IP for node %s: %w", nodeName, err)
+	}
+	return transitSwitchIP[0].IP.String(), nil
```
good catch from @kyrtapz : Forgot about dualstack? :D
LOL meant to get back to this but didn't. I guess this is the side effect of working on a single PR for 4 months, you start to loose context on what all you wanted to do.

tssurya commented on go-controller/pkg/clustermanager/clustermanager.go:
```diff
@@ -93,6 +98,12 @@ func (cm *ClusterManager) Start(ctx context.Context) error {
 	}
 
+	if config.OVNKubernetesFeature.EnableEgressIP {
+		if err := cm.eIPC.Start(); err != nil {
```
good catch!
I swear I added a nice `Stop` function that seems like dead code at this point :)
```
func (eIPC *egressIPController) Stop() {
	close(eIPC.stopChan)
	eIPC.wg.Wait()
	if eIPC.egressNodeHandler != nil {
		eIPC.watchFactory.RemoveNodeHandler(eIPC.egressNodeHandler)
	}
	if eIPC.egressIPHandler != nil {
		eIPC.watchFactory.RemoveEgressIPHandler(eIPC.egressIPHandler)
	}
	if eIPC.cloudPrivateIPConfigHandler != nil {
		eIPC.watchFactory.RemoveCloudPrivateIPConfigHandler(eIPC.cloudPrivateIPConfigHandler)
	}
}
```

tssurya replied:
done.

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -0,0 +1,1768 @@
+	objretry "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/retry"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	kapi "k8s.io/api/core/v1"
```
ack thanks for noticing that, bad VS code import logic. cleaned this up.

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -0,0 +1,1768 @@
+func (eIPC *egressIPController) initRetryFramework() {
+	eIPC.retryEgressNodes = eIPC.newRetryFramework(factory.EgressNodeType)
+	eIPC.retryEgressIPs = eIPC.newRetryFramework(factory.EgressIPType)
+	eIPC.retryCloudPrivateIPConfig = eIPC.newRetryFramework(factory.CloudPrivateIPConfigType)
```
hmm I copied from: https://github.com/ovn-org/ovn-kubernetes/blob/2254039c62f4aa3139c64586df0195354e687f40/go-controller/pkg/ovn/default_network_controller.go#L238
but I think the right thing to do is to check the util.Platform before initializing this, can save us some resources..

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -0,0 +1,1768 @@
+
+// AddResource adds the specified object to the cluster according to its type and
+// returns the error, if any, yielded during object creation.
+func (h *egressIPClusterControllerEventHandler) AddResource(obj interface{}, fromRetryLoop bool) error {
```
done.

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -0,0 +1,1768 @@
+	case factory.EgressNodeType:
+		oldNode := oldObj.(*kapi.Node)
+		newNode := newObj.(*kapi.Node)
+		// Initialize the allocator on every update,
```
yup, the change in nodeIP's is relevant only for the default LRPs we create for EIP at priority 102. Since that needs to be done from ovnkube-controller (previously master) we don't need this here.

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -0,0 +1,1768 @@
+}
+
+func (h *egressIPClusterControllerEventHandler) AreResourcesEqual(obj1, obj2 interface{}) (bool, error) {
+	return false, nil
```
yeah good point, I don't know why we always force an update in our current code: https://github.com/ovn-org/ovn-kubernetes/blob/2254039c62f4aa3139c64586df0195354e687f40/go-controller/pkg/ovn/base_event_handler.go#L95
I think we can save some cycles by doing a DeepEqual I guess. But on the other hand, if we do get an update event it means something changed :D in which case doing a deepequal does that really make sense?

tssurya replied:
ah wait for `EgressNode` type within the updateResource we hold logic for return.
for egressIP object type, we care both about `.Spec` changes and `.Status` changes so we won't say any cycles there. So leaving this as is for now.

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -0,0 +1,1768 @@
+	return assignableNodes, allAllocations
+}
+
+// setupNodeForEgress sets up default logical router policy for every node and
```
good point! done.

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -0,0 +1,1768 @@
+	eIPC.egressIPAssignmentMutex.Lock()
+	defer eIPC.egressIPAssignmentMutex.Unlock()
+
+	// Initialize an empty name which is filled depending on the operation
```
thanks for catching this.. done.

tssurya commented on go-controller/pkg/ovn/default_network_controller.go:
```diff
@@ -912,7 +887,9 @@ func (h *defaultNetworkControllerEventHandler) UpdateResource(oldObj, newObj int
-
+		// Update node in zone cache; value will be true if node is local
+		// to this zone and false if its not
+		h.oc.eIPC.nodeZoneState.Store(newNode.Name, h.oc.isLocalZoneNode(newNode))
```
So as we spoke offline on huddle, I have changed this up.

1. Now the new cache is done using our custom syncmap package.
2. All major paths are using the cache with a lock on the key: example: `addPodEgressIPAssignment`, `deletePodEgressIPAssignment` and `deleteEgressIPStatusSetup`.
3. I couldn't lock some parts like because it was tricky to do so :/ (open to ideas): 

A) In `deleteEgressIPStatusSetup` the part where we delete address-sets using `podIPs` is unsafe
```
// TODO(tssurya): Would be nice to lock the address-set deletion as well but its not that important
if len(podStatus.egressStatuses) == 0 && len(podStatus.standbyEgressIPNames) == 0 {
	// pod could be managed by more than one egressIP
	// so remove the podKey from cache only if we are sure
	// there are no more egressStatuses managing this pod
	klog.V(5).Infof("Deleting pod key %s from assignment cache", podKey)
	// delete the podIP from the global egressIP address set since its no longer managed by egressIPs
	if err := oc.deletePodIPsFromAddressSet(podIPs); err != nil {
		return fmt.Errorf("cannot delete egressPodIPs for the pod %s from the address set: err: %v", podKey, err)
	}
	delete(oc.eIPC.podAssignment, podKey)
} else if len(podStatus.egressStatuses) == 0 && len(podStatus.standbyEgressIPNames) > 0 {
	klog.V(2).Infof("Pod %s has standby egress IP %+v", podKey, podStatus.standbyEgressIPNames.UnsortedList())
	podStatus.egressIPName = "" // we have deleted the current egressIP that was managing the pod
	if err := oc.addStandByEgressIPAssignment(podKey, podStatus); err != nil {
		klog.Errorf("Adding standby egressIPs for pod %s with status %v failed: %v", podKey, podStatus, err)
		// We are not returning the error on purpose, this will be best effort without any retries because
		// retrying deleteEgressIPAssignments for original EIP because addStandByEgressIPAssignment failed is useless.
		// Since we delete statusToRemove from podstatus.egressStatuses the first time we call this function,
		// later when the operation is retried we will never continue down the loop
		// since we add SNAT to node only if this pod was managed by statusToRemove
		return nil
	}
}
```
B) In beginning of `addPodEgressIPAssignment` we use the cache to determine if we need to proceed with the ops:
```
// We need to proceed with add only under two conditions
// 1) egressNode present in at least one status is local to this zone
// (NOTE: The relation between egressIPName and nodeName is 1:1 i.e in the same object the given node will be present only in one status)
// 2) the pod being added is local to this zone
// TODO(tssurya): Check if this loop also needs to be locked for nodeZoneState key?
proceed := false
for _, status := range statusAssignments {
	isLocalZoneNode, loadedEgressNode := oc.eIPC.nodeZoneState.Load(status.Node)
	if loadedEgressNode && isLocalZoneNode {
		proceed = true
		break
	}
}
if !proceed && !oc.isPodScheduledinLocalZone(pod) {
	return nil // nothing to do if none of the status nodes are local to this master and pod is also remote
}
```

tssurya replied:
oh err and maybe I need to always lockKey before loading and unloading as well? not sure...I feel like since we hold lock on entire cache we should be safe?

tssurya replied:
resolving this comment, fixed this.

tssurya commented on go-controller/pkg/ovn/default_network_controller.go:
```diff
@@ -974,13 +939,12 @@ func (h *defaultNetworkControllerEventHandler) UpdateResource(oldObj, newObj int
 			}
-		} else if isNewReady && isNewReachable {
+		} else if isNewReady {
 			klog.Infof("Node: %s is ready and reachable, adding it for egress assignment", newNode.Name)
```
we won't know if its reachable or not unfortunately :/ before ovnkube-master used to check for reachable, ready and egressLabels, this behaviour will change to only checking for ready and egressLabels. We will add nodeIP to the address set of the 102 default LRP and do the GARP setup even if node is not reachable, although if node becomes non-reachable and if that node is used by egressIP, cluster-manager will trigger a status update and we will take action and delete that. That is our only signal for reachability. This was done intentionally because maintaining the whole healthcheck portions across all masters seemed like an overkill just to maintain that logic here.

As for nodes that are not allocated, tough luck.

tssurya replied:
@trozet and I spoke in depth about this and reached the conclusion that we don't really need any of the stuff we do in `reconcileNode`.
We will keep `GARP`, `addressSet` and `nodeZoneState` add/delete/update. no need for anything else..

trozet replied:
We talked on slack about removing most of reconcileNodeForEgressIP

tssurya replied:
done.

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -56,256 +49,109 @@ func getEgressIPAddrSetDbIDs(name egressIpAddrSetName, controller string) *libov
+		if err != nil {
+			return fmt.Errorf("invalid new podSelector, err: %v", err)
+		}
+		// CASE 3.2: Only Namespace selectors on Spec changed
```
yeah I had to write it down for my own sanity :D

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1101,26 +483,51 @@ func (oc *DefaultNetworkController) addPodEgressIPAssignments(name string, statu
+	// 2) the pod being added is local to this zone
+	proceed := false
+	for _, status := range statusAssignments {
+		isLocalZoneNode, loadedEgressNode := oc.eIPC.nodeZoneState.Load(status.Node)
```
hmm yeah I agree, check my comment: https://github.com/ovn-org/ovn-kubernetes/pull/3386#discussion_r1195712077 I tried to do a full lock but that was tricky across both parts. See my new push and if you have ideas on how I can include this loop as well as the add below nicely in our `DoWithLock` I can redo this also to be locked.

tssurya replied:
added locks before accessing the key here as discussed on slack. resolving this comment.

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1228,6 +608,14 @@ func (oc *DefaultNetworkController) deleteEgressIPAssignments(name string, statu
 				// we can continue here since this pod was not managed by this statusToRemove
 				continue
 			}
+			// this statusToRemove was managing at least one pod, hence let's tear down the setup for this status
```
yes, it was done because of your suggestion actually hehe.
Remember you, me and Pat had a call few weeks ago about a race bug? When we add things we look at the cache to see if it exists or not and add it accordingly, but when we delete things we weren't checking the cache. Hence we were removing setup for stale statuses which removed setup for other newly added statuses. So not totally for `if pod exists in cache` ; more for `does status exist in cache`.

We had so many stale NAT bugs that we clean too well :)
See
```
routerName := util.GetGatewayRouterFromNode(status.Node)
		natPred := func(nat *nbdb.NAT) bool {
			// TODO(tssurya): investigate if we should be deleting NATs only from status.Node
			return nat.ExternalIDs["name"] == name && nat.ExternalIP == status.EgressIP
		}
```
within `deleteEgressIPStatusSetup`
So we had two options, 1) either check cache before triggering delete OR 2) delete NAT based on status.Node - we were hesitant to do the first approach.

tssurya replied:
I am sometimes having a bad event race only in CI - its the same as we discussed here; https://github.com/ovn-org/ovn-kubernetes/pull/3386#discussion_r1196790084
EIP is assigned to node1:
```
2023-05-22T16:54:26.7175003Z I0522 16:54:26.717311   21445 egressip.go:1206] Patching status on EgressIP egressip: [{node1 192.168.126.101}]
```
Label changes to node2:
```
2023-05-22T16:54:26.7208352Z I0522 16:54:26.720620   21445 egressip.go:1206] Patching status on EgressIP egressip: [{node2 192.168.126.101}]
```
Add pod setup for node1 status:
```
2023-05-22T16:54:26.7314476Z I0522 16:54:26.731207   21445 egressip.go:574] Adding pod egress IP status: {node1 192.168.126.101} for EgressIP: egressip and pod: egressip-namespace/egress-pod/[10.128.0.15/23]
```
Add pod setup for node2 status:
```
2023-05-22T16:54:26.7698191Z I0522 16:54:26.769604   21445 egressip.go:574] Adding pod egress IP status: {node2 192.168.126.101} for EgressIP: egressip and pod: egressip-namespace/egress-pod/[10.128.0.15/23]
```
and finally the stupid delete for node1 status happens:
```
2023-05-22T16:54:26.7921426Z I0522 16:54:26.791895   21445 egressip.go:636] Deleting pod egress IP status: {node1 192.168.126.101} for EgressIP: egressip
```
and because of https://github.com/ovn-org/ovn-kubernetes/blob/0e395d5a831930fd9f5e91d3303bf6d636da6136/go-controller/pkg/ovn/egressip.go#L2620 where we don't check for node, we remove all the NATs of all the nodes - this is not a problem in IC world where we only have one node but its failing my UTs for multiple nodes in same zone its happened twice now:  https://github.com/ovn-org/ovn-kubernetes/actions/runs/5048097599/jobs/9055855508?pr=3386 and https://github.com/ovn-org/ovn-kubernetes/actions/runs/5048097599/jobs/9055855508?pr=3386. I am going to re-add the status.Node match to remove NATs that are only on the node we care about!

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1614,21 +1032,36 @@ func (oc *DefaultNetworkController) generateCacheForEgressIP() (map[string]egres
-		}
-		namespaces, err := oc.watchFactory.GetNamespacesBySelector(egressIP.Spec.NamespaceSelector)
-		if err != nil {
+			_, isLocalZoneNode := oc.localZoneNodes.Load(status.Node)
```
ah yeah good point. Let me use `isEgressNodeLocalZone`. I agree `isLocalZoneNode` is confusing between pod's node and egress node.

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1614,21 +1032,36 @@ func (oc *DefaultNetworkController) generateCacheForEgressIP() (map[string]egres
+				}
+				nextHopIP = gatewayRouterIP.String()
+			} else {
+				nextHopIP, err = oc.eIPC.getTransitSwitchIP(status.Node)
```
ah I like `getTransitIP` better, makes it feel like I am appariting (https://harrypotter.fandom.com/wiki/Apparition) like on harry potter from one zone (world) to another.

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -2288,29 +1431,41 @@ func (e *egressIPController) addPodEgressIPAssignment(egressIPName string, statu
+			return fmt.Errorf("unable to create NAT rule ops for status: %v, err: %v", status, err)
+		}
+		if config.OVNKubernetesFeature.EnableInterconnect && (loadedPodNode && !isLocalZonePod.(bool)) {
+			// configure reroute for non-local-zone pods on egress nodes
```
> i'm a little confused when we are adding static routes and policies. So if the egress node is local and the pod is not local, then we configure a static route on the pod's node?

correct.

> 
> The static route reroutes towards the GR IP of the egress node.

yes.

> 
> Then for the policy...if the pod node is local then we create a reroute policy?

yes.

> 
> The policy then:
> 
> 1. routes to GR IP if the egress node is local OR
> 2. routes to the transit router IP
> 
> Is that right? If so, could we have that in this function's comments?

yes you got all that right. See https://docs.google.com/presentation/d/1jdZrkFVTc9O9-d2I6SCa8EsBCaaxg2dD_x0868Izqdw/edit#slide=id.g229128c4b5c_0_5
I am actually doing a detailed egressIP on Interconnect documentation: https://github.com/ovn-org/ovn-kubernetes/pull/3526. This will all be outlined there. Is that ok or do you also want to see some in the code?

> 
> Is there a reason we use static routes for the reroute instead of policy? I was thinking that if we only used policy then we could use a unique id since only one should exist on a router per pod so we could never have a case where a static route and a policy were configured on the same node. We could use client side indexes to update the policy in case egress node moves and not have to worry about if we left some stale static route or policy around. wdyt?

We spoke about this in huddle, we can't use LRPs because LRSR is evaluated first in OVN pipeline and as soon as traffic is sent to the remote node, it get's dropped before even reaching LRP if we don't have the right LRSR's
As for the indexes part, I think +1 to that idea, but that is going to take considerable cycles, so as decided let's ensure to track a feature work or improvements card or something for this.

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -2448,19 +1633,31 @@ func ipFamilyName(isIPv6 bool) string {
+			return nil, fmt.Errorf("unable to retrieve gateway IP for node: %s, protocol is IPv6: %v, err: %w", status.Node, isEgressIPv6, err)
+		}
+		nextHopIP = gatewayRouterIP.String()
+	} else if config.OVNKubernetesFeature.EnableInterconnect {
```
yes actually this whole function will be called only if:
```
	// don't add a reroute policy if the egress node towards which we are adding this doesn't exist
	if loadedEgressNode && loadedPodNode && isLocalZonePod {
		ops, err = e.createReroutePolicyOps(ops, podIPs, status, egressIPName)
		if err != nil {
			return fmt.Errorf("unable to create logical router policy ops, err: %v", err)
		}
		ops, err = e.deleteExternalGWPodSNATOps(ops, pod, podIPs, status)
		if err != nil {
			return err
		}
	}
```
loadedEgressNode is true, so node needs to exist. I will add a comment to clarify this point.

tssurya replied:
In fact I should not check loadedEgressNode here again.

tssurya replied:
spoke with @trozet :
Better to do:
```
if nextHop == nil {
    return err, YIKES COULD NOT FIND IP!!!
}
```

trozet replied:
talked about just adding a check to make sure nextHopIP is not empty before we go to create the OVN NBDB ops

tssurya replied:
this is done.

tssurya commented on go-controller/pkg/ovn/egressip_test.go:
```diff
@@ -7,11 +7,11 @@ import (
 	"time"
 
 	"github.com/onsi/ginkgo"
+	ginkgotable "github.com/onsi/ginkgo/extensions/table"
```
yeah @flavio-fernandes went through this and I haven't reinvented the wheel, I just made the tests use various scenarios of interconnect like -> node1 and node2 in same zone, versus they are different zones and so on...
I don't anticipate any breaking changes to this UT file.

flavio-fernandes commented:
Update: this is a /lgtm... pending @trozet , @kyrtapz and others agree, ofc. ;)

tssurya commented on go-controller/pkg/ovn/default_network_controller.go:
```diff
@@ -772,7 +757,12 @@ func (h *defaultNetworkControllerEventHandler) AddResource(obj interface{}, from
+		// Update node in zone cache; value will be true if node is local
+		// to this zone and false if its not
+		h.oc.eIPC.nodeZoneState.LockKey(node.Name)
+		h.oc.eIPC.nodeZoneState.Store(node.Name, h.oc.isLocalZoneNode(node))
```
lock here is to protect the write we are doing into `nodeZoneState`

flavio-fernandes replied:
ack. Then my source of confusion is that SyncMaps have a lock and that should be enough.
But I can now see that this is not the case. Here. All good.

ref: https://github.com/ovn-org/ovn-kubernetes/pull/3097

```
// LockKey should be called before reading/writing entry value,
// it guarantees exclusive access to the key.
// Unlock(key) should be called once the work for this key is done to unlock other threads
// After the key is unlocked there are no guarantees for the entry for given key
func (c *SyncMap[T]) LockKey(key string) {
...

// Store sets the value for a key.
// If key-value was already present, it will be over-written
func (c *SyncMap[T]) Store(lockedKey string, newEntry T) {
...
```

flavio-fernandes commented:
/retest

tssurya commented:
Fixing the UTs for GARP -> since we will now always have them for all nodes...

tssurya commented:
> Fixing the UTs for GARP -> since we will now always have them for all nodes...

finally DONE!

tssurya commented:
note shard conformance is not really failing:
```
Ran 332 of 7069 Specs in 3620.611 seconds
FAIL! - Suite Timeout Elapsed -- 332 Passed | 0 Failed | 4 Flaked | 0 Pending | 6737 Skipped
```
it timed out will retest this once control plane's pass.

tssurya commented:
/retest-failed

tssurya commented:
Opened a new PR for moving the metrics over to cluster manager for 1) egressIPCount 2) egressIPReachability 3) egressIPRebalance : https://github.com/ovn-org/ovn-kubernetes/pull/3611
Once this goes in, we should get that one in as well.

flavio-fernandes commented:
🥳🎊🎉
